### ISSUE TITLE
twisted.web.client.readBody deprecation warning is too aggressive

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7762]
|<img alt="bdarnell's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| bdarnell reported|
|-|-|
|Trac ID|trac#7762|
|Type|release blocker: regression|
|Created|2015-01-30 16:46:00Z|
|Branch|https://github.com/twisted/twisted/tree/readbody-deprecation-spurious-7762|

The deprecation warning from twisted.web.client.readBody (new in 15.0: https://twistedmatrix.com/trac/changeset/43379) may fire inappropriately. It looks like a timing problem: when the response is small enough that the headers and body are read all at once, the transport is already set to None by the time readBody tests for the presence of the abortConnection method. This leads to spurious DeprecationWarnings (when those warnings are enabled).

<details><summary>Searchable metadata</summary>

```
trac-id__7762 7762
type__release_blocker__regression release blocker: regression
reporter__bdarnell bdarnell
priority__high high
milestone__None None
branch__branches_readbody_deprecation_spurious_7762 branches/readbody-deprecation-spurious-7762
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1422636360593804 1422636360593804
changetime__1445217222500211 1445217222500211
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

Thanks for reporting this!

I'll mark this for 15.1, to make sure we've got a solution before then, and to make sure it doesn't slip through the cracks.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

Hi bdarnell,

Thanks very much for filing this.  However, it's really unfortunate that we didn't catch this during the pre-release cycle.  Could we have communicated the prerelease in some other way which would have enabled you to test this in time?

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

(In [43885]) Branching to readbody-deprecation-spurious-7762.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl removed owner|
|-|-|

I've done up a failing test and a fix, so we can have this regression fixed and 15.1 getting rolling.

Putting up for review.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl commented|
|-|-|

(builders spun)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban commented|
|-|-|

Thanks for the code. It looks good.

Is this just a misc ticket? I think that is worth having a note in the release notes as at lease one person has observed :)

-----------

Minor comments.

The existing code already does this... but why not do the check using 

```
ITCPTransport.providedBy(protocol.transport)
```

since we are talking about an zope.interface, why check for duck typing using getattr abortConnection?


Why not check for deprecated condition error only in cancel method? in this way you will call getattr once.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set owner to @hawkowl|
|-|-|

Hi hawkie.  Thank you for this fix.  And thank you for spinning the builders, always nice to come to an all-green ticket just ready for review of the changes themselves :).

Normally I'd agree with adi that we should use a formally defined interface.  However, `abortConnection` is in a slightly awkward place; it lives on `ITCPTransport` but things that do not offer other TCP-like features really ought to have this method as well.  And some do, at least outside of Twisted.  Also, that's not really the subject of this change, so let's not feature-creep, especially on a regression.

 1. For now, let's stick with `getattr`/`abortConnection`, but please make the deprecation warning consistent with the actual implementation: `"Using readBody with a transport that does not haven an abortConnection method"`.  (I realize this is _also_ technically out of scope but hopefully this minor documentation update isn't too onerous.)
 2. More to the point of _this_ change, I think that `if protocol.transport:` ought to be `if protocol.transport is None:`.  There's no need to call the `__nonzero__` method on the transport - we just want to know if we have one or not.

Modulo these minor tweaks I think this is good to land.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl commented|
|-|-|

(In [43953]) review comments, refs [#7762](https://github.com/twisted/twisted/issues/7762)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

Thanks Glyph!

Spinning the builders, then will merge.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl set status to `closed`|
|-|-|

(In [43956]) Merging readbody-deprecation-spurious-7762: twisted.web.client.readBody deprecation warning is too aggressive

Author: hawkowl
Reviewers: adiroiban, glyph
Fixes: [#7762](https://github.com/twisted/twisted/issues/7762)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

> More to the point of this change, I think that if protocol.transport: ought to be if protocol.transport is None:. There's no need to call the __nonzero__ method on the transport - we just want to know if we have one or not.

It seems like this was turned in to `if not protocol.transport is None:` which .. works .. but it's probably the most confusing way to spell the check.

Also, unfortunately, the exact same mistake as was pointed out in the review comment appears just a few lines lower down and wasn't fixed - `if protocol.transport and protocolIsAbortable is None:`.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

Thanks for following up on the merge, exarkun.

I've addressed that feedback on [#7802](https://github.com/twisted/twisted/issues/7802), which is in review, with more or less green builds (one retest for an unrelated intermittent failure pending as of this writing).

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_13"></a>|@hawkowl commented|
|-|-|

Ticket retargeted after milestone deleted
